Make /var/cache writable and disable dynamic motd

woodrow-shen opened the pull request:
Bug discussion: https://github.com/snapcore/core18/issues/117

vorlonofportland commented on hooks/020-extra-files.chroot:
```diff
@@ -35,3 +35,6 @@ mkdir -p /var/lib/cloud
 ln -s ../lib/snapd/snapctl /usr/bin/snapctl
+
+echo "creating motd-news file"
+touch /var/cache/motd-news
```
I believe we need to make /var/cache generally writable for the core snap, not special-case /var/cache/motd-news.

woodrow-shen replied:
Agree +1

sil2100 commented:
This change looks good, but I do agree with Steve here - I guess the 'better' way would be to just make /var/cache writable. @mvo5 do you know of any reasons why we shouldn't do that?

mvo5 commented:
I think its fine to make /var/cache as a whole writable. The reason we did not so far is that we are quite targeted about writable-path handling historically. But there is no technical reason for this.

A bit of a meta-question - do we want dynamic motd on core18 at all? On core we only had a static motd. If we want a dynamic one, how does it update itself?

vorlonofportland commented:
> A bit of a meta-question - do we want dynamic motd on core18 at all? On core we only had a static motd. If we want a dynamic one, how does it update itself?

@mvo5 Indeed, that was a question I raised earlier on the linked issue.  We might in fact want to change BOTH things - make /var/cache writable, and disable update-motd in core.

The costs of not disabling update-motd in core are:
- on all systems, a query of the motd.ubuntu.com service once every 12h
- on systems where users log in, the scripts in /etc/update-motd.d/ will be run on each login.

Currently, the output of update-motd on core18 will not match the existing static /etc/motd and will not be correct for core systems, so we would need to deal with that.

woodrow-shen commented:
Hi @vorlonofportland and @mvo5 ,

I resent a commit based on above recommendations, and also I verified the test image with this patch. Please give some advice if any improvement, thanks.

vorlonofportland commented on hook-tests/014-set-motd.test:
```diff
@@ -7,3 +7,5 @@ grep -q "This Ubuntu Core 18 machine is a tiny" etc/motd
 test ! -e lib/systemd/system/motd-news.service
 test ! -e lib/systemd/system/motd-news.timer
+
+grep -q "ENABLE=0" etc/default/motd-news
```
This should still be ENABLED=0, let's not change the variable name

woodrow-shen replied:
Absolutely right, will fix it.

vorlonofportland commented on hooks/014-set-motd.chroot:
```diff
@@ -23,3 +23,6 @@ rm /etc/update-motd.d/10-help-text
 rm /lib/systemd/system/motd-news.{service,timer}
+
+# diable dynamic motd
+sed -i 's/ENABLED=1/ENABLE=0/g' /etc/default/motd-news
```
ENABLED=0, not ENABLE=0.

This makes sense, to avoid the network hit to poll for content that won't be displayed, as part of the console login.

woodrow-shen replied:
Oops, it's a ridiculous mistake, I'll fix it.

vorlonofportland commented on static/etc/system-image/writable-paths:
```diff
@@ -12,7 +12,7 @@
 /var/lib/snapd                         auto                     persistent  transition  none
 # snap catalogue (sections, pkgnames)
-/var/cache/snapd                        auto                    persistent  transition  none
+/var/cache                              auto                    persistent  transition  none
```
The expected behavior here on upgrade of the core18 snap is that both /writable/system-data/var/cache/snapd and /writable/system-data/var/cache/apparmor will already exist, and therefore any other contents of /var/cache on the read-only squashfs will be ignored; and /writable/system-data/var/cache will be mounted to /var/cache instead of mounting the individual subdirectories.

This looks reasonable and acceptable to me, as the only other contents of /var/cache in the current core18 snap are /var/cache/ldconfig (readable only by root and unused at runtime) and /var/cache/fontconfig (an empty directory).

woodrow-shen replied:
Thanks for bringing the upgrading concern, let's keep it to see if @mvo5 has any feedback to this.

Commit:
Make /var/cache writable and disable dynamic motd

Based on current discussions, we will intend to make /var/cache as a
whole writable, so redundant paths can be removed from writable-paths.
Meanwhile, there is a suggestion to disable dynamic motd until we have
better solution.

vorlonofportland commented:
I'm +1 on this and we have a verbal +1 from @mvo5 .
